src/cache/http_cache.go

```diff
@@ -18,178 +19,186 @@ import (
+	hdr.Gid = nobody
+	hdr.Uname = "nobody"
+	hdr.Gname = "nobody"
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	} else if info.IsDir() || target != "" {
+		return nil // nothing to write
+	}
+	f, err := os.Open(name)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = io.Copy(tw, f)
+	return err
+}
 
-	response, err := http.Get(cache.URL + "/artifact/" + artifact)
+func (cache *httpCache) Retrieve(target *core.BuildTarget, key []byte, files []string) *core.BuildMetadata {
+	m, err := cache.retrieve(target, key)
 	if err != nil {
-		return false
-	}
-	defer response.Body.Close()
-	if response.StatusCode == 404 {
-		return false
-	} else if response.StatusCode < 200 || response.StatusCode > 299 {
-		log.Warning("Error %d from http cache", response.StatusCode)
-		return false
-	} else if response.Header.Get("Content-Type") == "application/octet-stream" {
-		// Single artifact
-		return cache.writeFile(target, file, response.Body)
-	} else if _, params, err := mime.ParseMediaType(response.Header.Get("Content-Type")); err != nil {
-		log.Warning("Couldn't parse response: %s", err)
-		return false
-	} else {
-		// Directory, comes back in multipart
-		mr := multipart.NewReader(response.Body, params["boundary"])
-		for {
-			if part, err := mr.NextPart(); err == io.EOF {
-				return true
-			} else if err != nil {
-				log.Warning("Error reading multipart response: %s", err)
-				return false
-			} else if !cache.writeFile(target, part.FileName(), part) {
-				return false
-			}
-		}
+		log.Warning("%s: Failed to retrieve files from HTTP cache: %s", target.Label, err)
 	}
+	return m
 }
 
-func (cache *httpCache) writeFile(target *core.BuildTarget, file string, r io.Reader) bool {
-	outFile := path.Join(target.OutDir(), file)
-	if err := os.MkdirAll(path.Dir(outFile), core.DirPermissions); err != nil {
-		log.Errorf("Failed to create directory: %s", err)
-		return false
+func (cache *httpCache) retrieve(target *core.BuildTarget, key []byte) (*core.BuildMetadata, error) {
+	req, err := http.NewRequest(http.MethodGet, cache.makeURL(key), nil)
+	if err != nil {
+		return nil, err
 	}
-	f, err := os.OpenFile(outFile, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, target.OutMode())
+	resp, err := cache.client.Do(req)
 	if err != nil {
-		log.Errorf("Failed to open file: %s", err)
-		return false
+		return nil, err
+	} else if resp.StatusCode == http.StatusNotFound {
+		return nil, nil // doesn't exist - not an error
+	} else if resp.StatusCode != http.StatusOK {
+		b, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf("%s", string(b))
 	}
-	defer f.Close()
-	if _, err := io.Copy(f, r); err != nil {
-		log.Errorf("Failed to write file: %s", err)
-		return false
+	defer resp.Body.Close()
+	gzr, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer gzr.Close()
+	tr := tar.NewReader(gzr)
+	for {
+		hdr, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return loadPostBuildFile(target), nil
+			}
```

Comment:
There’s a `return` missing here. If `tr.Next` returns a non-`io.EOF` error, the line below will panic.

Reply:
Good spot, thank you!